# Fitbod EDA
**Author:** Liam Gundlach

## Goals
When digging through this data, my goals are to first find a measurement of success of a user and then to see what it is that fitbod does that plays a role in making a user successful. I aim to produce both potential action items as well as outlines for follow-up analyses that could produce further actionable insights.

In [47]:
# importing packages for data manipulation, plotting, and querying my local postgres db
import pandas as pd
from sqlalchemy import create_engine
# import psycopg2
import plotly.express as px
from sklearn.linear_model import LinearRegression
import sklearn.metrics as metrics
import statsmodels.formula.api as sm
from statistics import mean
import numpy as np
from joblib import Parallel, delayed
import scipy
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
engine = create_engine('postgresql://postgres:un1versaL!!!!@localhost:5432/postgres')

Reading in the data and placing it in a postgres table named `fitbod`

In [2]:
df = pd.read_csv('/Users/liamgundlach/Downloads/FitbodHW_Liam_Gundlach/fitbod_Liam_Gundlach_20220328-104458.csv')
df.columns = [col.lower() for col in df.columns]
df.to_sql('fitbod',con=engine,if_exists='replace')
df.head()

date   workoutid  exerciseid singlesetid  \
0  2021-07-25  LjqB1qwZ5z  nNMm3ywt6H  pQNtY3J2x9   
1  2021-07-10  XKCH2FLX8c  oCzmU1KcYe  YoyGYzeaHr   
2  2021-12-13  OENRUB81i2  otjt26R0oL  HL8e8czc1p   
3  2021-03-15  N6tS6dVIvc  k3yxx2a6oz  Y6ItzSIr1g   
4  2021-08-15  GjzV186eNg  oDI8CKhqJi  YNsPoihUAR   

                       exercisename  reps     weight fake_user_id  
0                      Pike Push Up    10   0.000000  fake_id_438  
1  Cable Shoulder Internal Rotation    25   6.803894  fake_id_434  
2            Barbell Shoulder Press     8  29.483539  fake_id_436  
3         Seated Machine Calf Press    25  31.751504  fake_id_434  
4          Single Arm Preacher Curl    10  13.607787  fake_id_434

## Discussion

The two most easily measureable components of success here seem to be number of workouts and strength progress. While strength progress is a helpful barometer for how much fitbod is providing added value to a user, not every user wants to simply build strength. Additionally, workout volume is more of a barometer for the success of fitbod in forming an exercise habit with a user, which is a building block for whatever fitness goal a user has. For this reason, I will primarily look to locate factors that play into workout frequency and volume while digging into this data.

Ideally I would love to be able to see a few other measures of user behavior in this data set. I am sure there are more that would be helpful, but some that come to mind are:

- workout number
- user workout changes
    - in-workout exercise swaps
    - workout refreshes 
    - preference alterations
- user attributes
    - age
    - gender
    - height
    - weight
    - lifting experience
    - workout goals
    - equipment and muscle group preferences

Workout number would be a particularly helpful stat in this dataset, as it would help to isolate users who were trying out fitbod for the first time. 

## Next Steps

Now, back to the data at hand...

I checked out the dates of this dataset to see if they were windowed in some way, turns out the data set is from all of 2021. The narrow time-scope of the data will make finding a stable panel of users a bit of a fools errand. (this is further confounded by my finding in a second that the user set has 71 users in it)

In [3]:
print(df['date'].min())
print(df['date'].max())

2021-01-01
2021-12-31


Checked out user stats to see what kind of range of users we are working with is and saw a distribution ranging from rare users to superusers. (output cut to head for readability here)

In [4]:
user_stats = pd.read_sql('''
select fake_user_id,
	   count(distinct workoutid) as total_workouts, 
	   count(distinct exerciseid) as unique_exercises,
	   count(distinct singlesetid) as total_sets,
	   sum(reps) as total_reps,
	   sum(weight*reps) as total_weight_lifted,
	   avg(weight) as avg_weight_lifted,
	   min(date(date)) as first_workout_date,
	   max(date(date)) as last_workout_date
from fitbod 
group by 1 
order by 2 desc
''', con = engine)
user_stats.to_sql('fitbod_user_stats',con=engine,if_exists='replace')

user_stats

fake_user_id  total_workouts  unique_exercises  total_sets  total_reps  \
0   fake_id_434             301                40        2725     32403.0   
1   fake_id_435             251                24        2768     20597.0   
2   fake_id_453             234                55        1924     25618.0   
3   fake_id_462             224                37        1632     19483.0   
4   fake_id_448             177                43         876     10356.0   
5   fake_id_433             173                49        2236     15063.0   
6   fake_id_438             152                53        1089     12591.0   
7   fake_id_471             135                27        1072     11613.0   
8   fake_id_461             134                46         792      8061.0   
9   fake_id_436             134                39         796      6647.0   
10  fake_id_491             131                18         982      8072.0   
11  fake_id_498             124                13         606      5664.0   
12  fake_id_467             121                15         628      7523.0   
13  fake_id_456             120                27         653      6912.0   
14  fake_id_459             118                36         829      8638.0   
15  fake_id_439             111                42         930     11149.0   
16  fake_id_460             109                30         703      7664.0   
17  fake_id_454             103                52         933     12037.0   
18  fake_id_442              99                42         603      5437.0   
19  fake_id_466              96                39         468      5002.0   
20  fake_id_463              93                41         740      6527.0   
21  fake_id_472              90                27         454      3702.0   
22  fake_id_451              83                41         635      5898.0   
23  fake_id_457              79                40         554      6014.0   
24  fake_id_465              73                40         413      4368.0   
25  fake_id_476              71                31         389      4496.0   
26  fake_id_479              71                35         351      3577.0   
27  fake_id_475              70                34         667      7796.0   
28  fake_id_450              69                31         420      4087.0   
29  fake_id_446              67                41         446      4663.0   
30  fake_id_443              67                17         379      4799.0   
31  fake_id_464              66                31         675      6514.0   
32  fake_id_487              62                21         347      3442.0   
33  fake_id_455              62                27         313      2873.0   
34  fake_id_440              60                30         300      3339.0   
35  fake_id_484              59                25         330      2791.0   
36  fake_id_492              55                14         303      2514.0   
37  fake_id_458              55                38         436      4008.0   
38  fake_id_478              51                22         251      3196.0   
39  fake_id_445              51                20         287      3229.0   
40  fake_id_486              48                24         310      3060.0   
41  fake_id_489              48                37         299      3019.0   
42  fake_id_470              47                39         290      3020.0   
43  fake_id_480              46                22         224      2533.0   
44  fake_id_444              45                33         354      4328.0   
45  fake_id_496              42                 9         321      2983.0   
46  fake_id_452              37                28         182      1756.0   
47  fake_id_468              35                22         191      1550.0   
48  fake_id_441              33                13         233      3794.0   
49  fake_id_477              32                13         134      1307.0   
50  fake_id_494              29                25         176      1689.0   

I also looked at what the "average" user looked like in terms of their usage. I looked at both median and mean. The mean user had more workouts, more sets, more reps, and lifted heavier on average than the median user, indicating that the superusers at the top of the distribution were likely to be skewing the distribution. 

In [5]:
user_stats[['total_workouts','unique_exercises','total_sets','total_reps','total_weight_lifted','avg_weight_lifted']].median()

total_workouts            59.000000
unique_exercises          27.000000
total_sets               330.000000
total_reps              3442.000000
total_weight_lifted    44370.458401
avg_weight_lifted         16.591743
dtype: float64

In [6]:
user_stats[['total_workouts','unique_exercises','total_sets','total_reps','total_weight_lifted','avg_weight_lifted']].mean()

total_workouts             72.521127
unique_exercises           26.577465
total_sets                513.507042
total_reps               5285.492958
total_weight_lifted    116698.245726
avg_weight_lifted          20.461612
dtype: float64

What was perhaps most interesting about the mean/median comparison was the one metric that didn't change much: **unique exercises**. The miniscule difference between the unique exercises in the median and average users suggests that exercise variety does not scale with volume for superusers. 

What this difference could imply is that there is a difference between the exercises that superusers employ and those employed by less frequent users. I wanted to investigate this difference with an eye towards seeing what drives users to work our more often. 

I began by looking at some summarry statistics of different exercises. I wanted to capture the relative frequency that users employed exercises, so I was particularly interested in the number of workouts per user.

In [7]:
exercise_stats = pd.read_sql('''
select exerciseid,
       exercisename,
	   count(distinct workoutid) as total_workouts, 
	   count(distinct fake_user_id) as unique_users,
	   count(distinct singlesetid) as total_sets,
	   count(distinct workoutid)*1.0/count(distinct fake_user_id) as workouts_per_user,
	   avg(distinct b.total_workouts) as avg_user_workout_count
from fitbod a
left join fitbod_user_stats b using (fake_user_id)
group by 1,2
order by 6
''', con = engine)
print((exercise_stats.head(20)['avg_user_workout_count']*exercise_stats.head(20)['unique_users']).sum()*1.0/exercise_stats.head(20)['unique_users'].sum())
print(exercise_stats.head(20)['avg_user_workout_count'].std()/20**.5)
exercise_stats.head(20)

101.98780487804878
8.744567807893574


exerciseid                exercisename  total_workouts  unique_users  \
0   9zYC4cN1jr  Handle Band Squat to Press               2             2   
1   SGoBKeSD4r        TRX Single Leg Squat               1             1   
2   pGKwaDfr6n       Glute Ham Raise Situp               1             1   
3   wwpzETMErU      Machine Shoulder Shrug               1             1   
4   iQzrqsifIr             Lateral Step Up               9             8   
5   PZ2ogZtiY4            TRX Pulse Lunges               5             4   
6   tSjey8z21h             TRX Chest Press               4             3   
7   JZcl0pTDEj            TRX Glute Bridge               8             6   
8   GbK7ZSbRC6   Kettlebell Sumo High Pull               7             5   
9   mXEVvIJ5nf      Standing Hip Adduction               9             6   
10  j9nla5kTfn       Kettlebell Hang Clean               3             2   
11  6wceEWn4V1         Incline Svend Press               3             2   
12  yR61QnjQAO       Medicine Ball Push Up              14             9   
13  Fa3azBAbyn                    Abs Bike              11             7   
14  r7LqFtCsNv           Decline Leg Raise               5             3   
15  ISJBgnjnsY              Single Arm Row               7             4   
16  vWKV6qv8Pv              Clean and Jerk               7             4   
17  LDHzRzwxaI       Exercise Ball Push Up              11             6   
18  O0VUqHDOIO            TRX Swimmer Pull              11             6   
19  LaUm5Jl8hd                   Sled Pull               4             2   

    total_sets  workouts_per_user  avg_user_workout_count  
0            5           1.000000               28.000000  
1            3           1.000000              173.000000  
2            3           1.000000              152.000000  
3            3           1.000000               55.000000  
4           36           1.125000               82.000000  
5           14           1.250000               84.500000  
6           18           1.333333              123.000000  
7           30           1.333333              125.833333  
8           28           1.400000              120.200000  
9           28           1.500000               87.666667  
10          11           1.500000              138.000000  
11          10           1.500000               89.000000  
12          52           1.555556              109.111111  
13          43           1.571429              113.000000  
14          18           1.666667              186.333333  
15          19           1.750000              111.000000  
16          24           1.750000              111.000000  
17          41           1.833333               65.666667  
18          42           1.833333               62.833333  
19          14           2.000000              118.500000

It looks like a lot of these workouts are of the bodyweight variety but some kettlebell, dumbell, and barbell exercises make it tough to get anything concrete here. 

Overall these definitely look the part of less frequently seen workouts if you are strolling around a gym but there isn't a ton to be gained here otherwise.

In [8]:

print((exercise_stats.sort_values('workouts_per_user',ascending=False).head(20)['avg_user_workout_count']*exercise_stats.sort_values('workouts_per_user',ascending=False).head(20)['unique_users']).sum()*1.0/exercise_stats.sort_values('workouts_per_user',ascending=False).head(20)['unique_users'].sum())
print(exercise_stats.sort_values('workouts_per_user',ascending=False).head(20)['avg_user_workout_count'].std()/20**.5)
exercise_stats.sort_values('workouts_per_user',ascending=False).head(20)

97.79946351847416
5.463905537923173


exerciseid                      exercisename  total_workouts  \
101  66ykNaEdxx       Smith Machine Decline Press              50   
100  UThEK7cf4f                       Machine Fly             429   
99   UQBXLgsfXN                    Dumbbell Shrug             321   
98   HggeHLeipS           Cable Double Bicep Curl             118   
97   oCzmU1KcYe  Cable Shoulder Internal Rotation              63   
96   1mSqEHZVVv               Machine Reverse Fly             156   
95   G1mxtZtEs5                        Back Squat             442   
94   516XLLabVp                  Scapular Pull Up              72   
93   k3yxx2a6oz         Seated Machine Calf Press             203   
92   ACSgVaipTW            Dumbbell Squeeze Press             292   
91   243X43p1XP                   Seated Leg Curl             343   
90   QUvE80fdlJ                   Tricep Kickback              54   
89   3GvUTFN8up                        Calf Raise             184   
88   zJlvUiMLou       Barbell Incline Bench Press             266   
87   Mok6zwcCl9                 Cable Upright Row             102   
86   otjt26R0oL            Barbell Shoulder Press             313   
85   9q2Cdgm0FM                 Machine Leg Press             218   
84   xPS7RzoOGm             Bent Over Barbell Row             240   
83   EAjKaB5zWy                Dumbbell Kickbacks             318   
82   djuvlRHUvb          Dumbbell Rear Delt Raise             317   

     unique_users  total_sets  workouts_per_user  avg_user_workout_count  
101             4         180          12.500000              148.000000  
100            35        1616          12.257143              103.103448  
99             28        1266          11.464286              111.642857  
98             11         479          10.727273              140.818182  
97              6         241          10.500000              164.000000  
96             15         680          10.400000              124.000000  
95             43        1766          10.279070               80.769231  
94              8         198           9.000000              136.000000  
93             23         827           8.826087              104.130435  
92             34        1054           8.588235               93.058824  
91             42        1294           8.166667               93.342105  
90              7         196           7.714286              117.000000  
89             24         718           7.666667              102.416667  
88             36        1059           7.388889               80.852941  
87             14         381           7.285714              136.500000  
86             44        1165           7.113636               92.775000  
85             31         805           7.032258              103.666667  
84             35         916           6.857143               89.030303  
83             50        1083           6.360000               82.377778  
82             50        1131           6.340000               88.444444

## Discussion

This looks like it could be a laundry list of exercises you are most likely to have to wait in line to do at they gym. The list of high-frequency exercises is therefot not anything groundbreaking here, but it does tell us that fitbod users tend to workout using a set of exercises that fall in line with popularly employed exercises. 

The high usage rate workouts did register a slightly lower average number of workouts than the bottom, but the difference doesn't apprear significant on the surface. Even without initial significance here in this small sample, this relationship could be worth looking into because of the potential implications.

When trying to parse out possible causality in this data, it is difficult to sort out the direction of causality in these relationships at a summary level. It would be great if this relationship was as simple as "give users less frequently employed exercises and they will work out more" but that is far from the only potential explanation for this relationship. In these measurements, it is likely that users who work out more often are more likely to recieve a higher variety of exercises, and you may even use them as test subjects for net new workouts becuase they are less likely to be lost as customers from it. Either of those factors (or many potential others) could skew the data to present correlations that look falsely masquarade as actionable causal relationships. 

## Next Steps

I want to test whether there is a relationship between this measure of exercise rarity and total usage of fitbod. To do this, I am going to seperate the users into two groups, using the first to measure exercise rarity and the second to match that exercise rarity to total workout counts. The hope here is that uncoupling the two measures will help to eliminate some of those supurfluous correlations.

Because the data set is so small, the division between the two user sets could swing the outcome wildly becuase of how much noise is likely to be contained in even smaller user samples. To circumvent this problem, I made a quick bootstrapping function to iterate over random divisions and accumulate the outputs into a list for visualization. This output list will give us an idea of the probability distribution of the true relationship between the two measurements. 


In [29]:


def bootstrap_comparison(df,fx,iterations=1000,n_jobs = 4, parallel = True):
    '''
    inputs:
    fx: function of groups a and b that compares the two groups and outputs a comparitive measure using the two groups
    iterations: iterations of the bootstrapping method's devision of the data into two groups
    '''
    if parallel == True:
        results = Parallel(n_jobs=n_jobs)(delayed(fx)(df) for _ in range(iterations))
    else:
        results = []
        for _ in range(iterations):
            iteration = fx(df)
            if type(iteration) == list:
                results += iteration
            else:
                results.append(iteration)
    return results

def exercise_usage_rate_to_workout_count(df):
    '''
    takes in a dataframe and two groups of users
    returns the ratio of total workouts between the users who were given "unpopular" workouts to those given "popular" workouts. (users can belong to both groups)
    the popular and unpopular workouts are determined from group a, where popularity is approximated by the number of workouts per user of the workout
    '''
    excercise_comparison = pd.read_sql('''
    with user_split as (
        select fake_user_id, row_number() over (order by random())*1.0/count(*) over () rand
        from (
            select distinct fake_user_id
            from fitbod
        ) a),

    exercise_popularity as(
        select exerciseid,
            exercisename,
            count(distinct workoutid) as total_workouts, 
            count(distinct fake_user_id) as unique_exercises,
            count(distinct workoutid)*1.0/count(distinct fake_user_id) as usage_rate,
            ntile(5) over (order by count(distinct workoutid)*1.0/count(distinct fake_user_id)) as usage_ntile
        from fitbod a
        left join user_split b using (fake_user_id)
        where rand <= .5
        group by 1,2
        order by 5 desc)

    select usage_rate,
            exerciseid,
            avg(d.total_workouts) as avg_total_workouts
    from fitbod a
    left join exercise_popularity b using (exerciseid,exercisename)
    left join user_split c using (fake_user_id)
    left join fitbod_user_stats d using (fake_user_id)
    where rand > .5
    group by 1,2

    ''',con=engine)
    # print(excercise_comparison.iat[0,0])
    ols = sm.ols(formula = 'avg_total_workouts ~ usage_rate', data=excercise_comparison)
    r = ols.fit()
    # print(excercise_comparison.iat[0,0])
    return {'intercept':r.params[0],'usage_rate_coef':r.params[1]}
    # return excercise_comparison.iat[0,1]/excercise_comparison.iat[1,1]

workout_count_projection = bootstrap_comparison(df,exercise_usage_rate_to_workout_count,iterations=2000,parallel=False)


In [30]:
workout_count_prediction_df = pd.DataFrame(workout_count_projection)
fig = px.histogram(pd.DataFrame(workout_count_prediction_df),x='usage_rate_coef')
mean_v1 = workout_count_prediction_df['usage_rate_coef'].mean()
print(f'The mean exercise usage rate coeficient is {mean_v1} in predicting total user workouts')

prob_greater_than_0 = workout_count_prediction_df.loc[workout_count_prediction_df['usage_rate_coef']>0]['usage_rate_coef'].count()*1.0/workout_count_prediction_df['usage_rate_coef'].count()
print(f'There is an approximately a {prob_greater_than_0} probability that exercise usage rate has a positive correlation with total user workouts')

fig.show()

The mean exercise usage rate coeficient is 3.5896318808635828 in predicting total user workouts
There is an approximately a 0.94 probability that exercise usage rate has a positive correlation with total user workouts


## Discusion

The usage rate measurement of these exercises in one group is fairly predictive of total workout count in the other, with a probability of positive relationship of 94%. This relationship indicates that something measured by exercise usage rate is correlated with user engagement, but lets dig into the possible causal implications of this correlation.

This trend could be indicative that prompting a user to perform a more popular exercise is more likely to promote future use of fitbod than prompting them to perform an unpopular exercise, but that is not neccessarily the case. A perhaps more likely hypothesis is that the popularity measurement is capturing something about structural differences between user types. Given earlier observations that unpopular exercises tend to be bodyweight exercises, and bodyweight exercises can be performed at home, it is possible that the difference observed here is due to at-home exercisers being less frequent in their workouts than gym-goers. This relationship could also suggest the existence of a filter where users who ultimately use fitbod a lot tend to only be ones that already enjoy popular exercises. 

On top of the user-caused posibilities here, there is also the possibility of a relationship inherent to the measurement itself, where higher usage rate exercises are higher usage because users performed them multiple times, and users who are more likely to perform them multiple times are those who have a higher workout count. This assumes some level of consistency across user groups, but I would bet on that consistency being present given that the algorithm recommending exercises is not going to be inherently different even if it will adapt to user preference.

## Next Steps

The most difficult aspect of learning from these results proved to be seperating user attributes from potential intervening action. For this reason, my next aim is to create a new measurement that represents fitbod's success relative to the individual user. 

To isolate this impact of fitbod workout curation, I am going to look at the relative time between workouts as a target metric. The relative time is the time between a workout and the next workout divided by the average time between workouts for that user. The assumption here is that, the more a user enjoys a fitbod workout, the sooner they will choose to perform a fitbod workout again. By normalizing to the frequency of an individual user's workouts, we can generalize the measurement across users with vastly different activity profiles that would otherwise have drastically different workout gaps. 

Ideally in most cases it would be great to be able to subdivide the users into multiple groups with more closely matching needs, but in a small sample like this neccessitates trying to use all the information on hand to tease out relationships between measurements in the data.

The first step here is to calculate the relative workout gaps of users, and then I will be using the exercise usage rate metric to attempt to predict this generalized measure of goodness of a workout. 

In [86]:
workout_gaps = pd.read_sql('''
with workout_gap as (
	select *,
	       lead(workout_date,1) over (partition by fake_user_id order by workout_date) as next_workout_date
	from
	(select distinct fake_user_id,
		   workoutid,
	 	   date(date) as workout_date
	from fitbod) a)
Select *,
       time_till_next_workout*1.0/avg(time_till_next_workout) over (partition by fake_user_id) as relative_workout_gap
from 
(select fake_user_id,
	   workoutid,
	   case when next_workout_date - workout_date = 0 then .5 
       else next_workout_date - workout_date end as time_till_next_workout
from workout_gap 
where next_workout_date is not null) a

''',con=engine)
workout_gaps.to_sql('workout_gaps',con=engine,if_exists='replace')

def exercise_use_rate_to_workout_gap(df):
    '''
    takes in a dataframe and two groups of users
    returns the ratio of total workouts between the users who were given "unpopular" workouts to those given "popular" workouts. (users can belong to both groups)
    the popular and unpopular workouts are determined from group a, where popularity is approximated by the number of workouts per user of the workout
    '''
    excercise_comparison = pd.read_sql('''
    with user_split as (
        select fake_user_id, row_number() over (order by random())*1.0/count(*) over () rand
        from (
            select distinct fake_user_id
            from fitbod
        ) a),

    exercise_popularity as(
        select exerciseid,
            exercisename,
            count(distinct workoutid) as total_workouts, 
            count(distinct fake_user_id) as unique_exercises,
            count(distinct workoutid)*1.0/count(distinct fake_user_id) as usage_rate
        from fitbod a
        left join user_split b using (fake_user_id)
        where rand <= .5
        group by 1,2
        order by 5 desc)

    select exerciseid,
           fake_user_id,
           usage_rate,
           avg(log(relative_workout_gap)) as user_relative_workout_gap
    
    from (
    select distinct 
        usage_rate,
        exerciseid,
        fake_user_id,
        relative_workout_gap
    from fitbod a
    left join exercise_popularity b using (exerciseid,exercisename)
    left join user_split c using (fake_user_id)
    left join workout_gaps d using (fake_user_id,workoutid)
    where rand > .5) a
    group by 1,2,3 having avg(log(relative_workout_gap)) between -100 and 100
    ''',con=engine)
    
    ols = sm.ols(formula = 'user_relative_workout_gap ~ usage_rate', data=excercise_comparison)
    r = ols.fit()
    return {'intercept':r.params[0],'usage_rate_coef':r.params[1]}

popularity_to_workout_frequency = bootstrap_comparison(df,exercise_use_rate_to_workout_gap,iterations=2000,parallel=False)



In [87]:
workout_gap_prediction_df = pd.DataFrame(popularity_to_workout_frequency)
fig = px.histogram(pd.DataFrame(workout_gap_prediction_df),x='usage_rate_coef')
mean_coef = workout_gap_prediction_df['usage_rate_coef'].mean()
mean_int = workout_gap_prediction_df['intercept'].mean()
print(f'The mean linear regression between usage rate and log relative workout gap has a coeficient of {mean_coef} and an intercept of {mean_int}')

prob_greater_than_0 = workout_gap_prediction_df.loc[workout_gap_prediction_df['usage_rate_coef']>0]['usage_rate_coef'].count()*1.0/workout_gap_prediction_df['usage_rate_coef'].count()
print(f'There is an approximately a {prob_greater_than_0} probability that exercise usage rate has a positive correlation with subsequent relative workout gap')

fig.show()

The mean linear regression between usage rate and log relative workout gap has a coeficient of 0.0006324801816318781 and an intercept of -0.1671552606740914
There is an approximately a 0.6235 probability that exercise usage rate has a positive correlation with subsequent relative workout gap


## Discussion

I will not be reading too much into a 63.25% chance of a positive relationship, but the direction of this slight relationship is certainly unexpected. The leaning of this trend suggests that it is likely exercises with large amounts of workouts per user are either correlated with increased relative workout gaps or are not correlated with them at all. This suggests that infrequently used exercises could push users to come back sooner to workout or at least do not push them to work out less frequently. It therfor seems increasingly likely that the exercise usage rate measurement is more a measure of algorithmic rarity rather than users choosing not to participate in the exercise.

Some contrasting hypotheses immediately jump to mind concerning this relationship. First is that prompting a user to perform a more uncommon exercise adds some excitement and variety to working out and prompts a user to return for their next workout quicker. Second is that some of these exercises are less common in the data because they provide less bang for your buck in terms of muscle exhaustion and are therefor chosen less often to populate users' workouts. This second hypothesis is in line with the more frequented exercises consisting of many compound movements and free weights while more uncommon exercises were more frequently bodyweight. 

## Next Steps

One of the things that immediately jumps to mind with the second hypothesis is that, if there is something about some exercises that is inherently less tiring, then we should see that variable be consistent across users and be reflective of similar subjective trends. Lets begin diving into this hypothesis by directly measuring the relative workout gap averages on an exercise level, as that an exercise-level variable produced this affect.

In [89]:
exercise_stats = pd.read_sql('''
select exerciseid,
        exercisename,
        avg(relative_workout_gap) as avg_relative_workout_gap,
        count(distinct workoutid) as total_workouts, 
        count(distinct fake_user_id) as unique_users,
        count(distinct workoutid)*1.0/count(distinct fake_user_id) as workouts_per_user
 from   (Select distinct exerciseid,
                exercisename,
                relative_workout_gap,
                workoutid,
                fake_user_id      
         from fitbod a
         left join workout_gaps d using (fake_user_id,workoutid)) a
group by 1,2 order by 3
''', con=engine)
exercise_stats.to_sql('fitbod_exercise_stats',con=engine,if_exists='replace')
exercise_stats

exerciseid                         exercisename  \
0    wwpzETMErU               Machine Shoulder Shrug   
1    mXEVvIJ5nf               Standing Hip Adduction   
2    6wceEWn4V1                  Incline Svend Press   
3    j9nla5kTfn                Kettlebell Hang Clean   
4    163u4e1Em9                 Hammerstrength Shrug   
5    ISJBgnjnsY                       Single Arm Row   
6    GbK7ZSbRC6            Kettlebell Sumo High Pull   
7    JyjzR9GyUB                             TRX Pike   
8    Fa3azBAbyn                             Abs Bike   
9    olwH9a4pKv                           Knee Raise   
10   6wga8VycHO                       Clean Deadlift   
11   pM3aRwuufC        Stability Ball Hyperextension   
12   O0VUqHDOIO                     TRX Swimmer Pull   
13   Yw1aA3soIp                  Dumbbell Sumo Squat   
14   r7LqFtCsNv                    Decline Leg Raise   
15   AOQBAQ8rXC                          Pendlay Row   
16   iQzrqsifIr                      Lateral Step Up   
17   crpeAKEXTd                   TRX Atomic Push Up   
18   y2Gg0PEzd8                        Preacher Curl   
19   QUvE80fdlJ                      Tricep Kickback   
20   3GvUTFN8up                           Calf Raise   
21   C4tOdXyE0R                  Cable Hip Abduction   
22   wNImc4b4MY                             Leg Curl   
23   qfkBJ4A5eQ                             Dead Bug   
24   1sQqqUqJyp                            Frog Pump   
25   jgnSb5e39S                   Australian Chin Up   
26   so3n0BqhED             Smith Machine Calf Raise   
27   XlazCsjHwn                          Hang Snatch   
28   ewEcdHovom                    Hanging Leg Raise   
29   yR61QnjQAO                Medicine Ball Push Up   
30   BN1DjSE1Zi                    Romanian Deadlift   
31   58REE5sJU1                        Russian Twist   
32   erjFOF21dc                 Reverse Barbell Curl   
33   QFhUumXlvb   Reverse Grip Barbell Bent Over Row   
34   243X43p1XP                      Seated Leg Curl   
35   9fptvki4pW                   Barbell Hip Thrust   
36   7wof3tzAVY  Smith Machine Stiff-Legged Deadlift   
37   zJlvUiMLou          Barbell Incline Bench Press   
38   k3yxx2a6oz            Seated Machine Calf Press   
39   06kzIMnTzR                       Oblique Crunch   
40   nNMm3ywt6H                         Pike Push Up   
41   516XLLabVp                     Scapular Pull Up   
42   LDHzRzwxaI                Exercise Ball Push Up   
43   otjt26R0oL               Barbell Shoulder Press   
44   aIzmfLklht                     Dumbbell Step Up   
45   gALxFFZcI4         Dumbbell Decline Bench Press   
46   Mok6zwcCl9                    Cable Upright Row   
47   uHBREgyju0                              Step Up   
48   k4ivWt2zuv                 Barbell Glute Bridge   
49   oDI8CKhqJi             Single Arm Preacher Curl   
50   9q2Cdgm0FM                    Machine Leg Press   
51   HpJxPvuCBP                       Dumbbell Clean   
52   G1mxtZtEs5                           Back Squat   
53   UThEK7cf4f                          Machine Fly   
54   DhkjvJNsNR                               Sit Up   
55   DdtF1OFcuQ                       Tricep Push Up   
56   UQBXLgsfXN                       Dumbbell Shrug   
57   r4bOO0fDTv                                Plank   
58   WoEeZr6Uwx              Single Arm Lat Pulldown   
59   mIMgSxPI0T                          Toes to Bar   
60   ACSgVaipTW               Dumbbell Squeeze Press   
61   BTmBzxhAmZ                    Dumbbell Pullover   
62   fxXbcxfkLt                 Exercise Ball Crunch   
63   dFycDkJyHe              Dumbbell Shoulder Raise   
64   4IMiKZHvCG                          Shotgun Row   
65   5hsZsuO5vt                Single Leg Hip Thrust   
66   JZcl0pTDEj                     TRX Glute Bridge   
67   9VAkz3bNl4                      Decline Push Up   
68   HggeHLeipS              Cable Double Bicep Curl   
69   EAjKaB5zWy                   Dumbbell Kickbacks   
70   KVt5QRq9Hx         Side 

Perusing the list of exercises ranked from smallest relative gap to largest, it would seem that the muscle exhaustion caused by the exercises is indeed playing a role in the frequency of workouts. Intensive compound movements like hang clean, lunge jump, and kettlebell upright row along with stability requiring movements like single arm overhead press, incline dumbbell row litter the top of the list, while isolation exercises like single leg raise, shrugs, and calf raises populate the bottom.

While this seems logical on the surface, I am going to check the data to see if these kinds of rules have a tendency to hold up across users. This approach will have the added benefit of seeing if measuring relative workout gap on an exercise level could be a useful measure for the postive impacts of individual workouts. 

As before, I am going to repeatedly split the data into a input measurement group and an output measurement set and run an OLS linear regression between the two measurements, saving the coeficient of the regression representing the translation of the input to output measurements.

In [82]:
def project_exercise_workout_gap(df):
    '''
    takes in a dataframe and two groups of users
    returns the ratio of total workouts between the users who were given "unpopular" workouts to those given "popular" workouts. (users can belong to both groups)
    the popular and unpopular workouts are determined from group a, where popularity is approximated by the number of workouts per user of the workout
    '''
    exercise_comparison = pd.read_sql('''
    with user_split as (
        select fake_user_id, row_number() over (order by random())*1.0/count(*) over () rand
        from (
            select distinct fake_user_id
            from fitbod
        ) a),

    exercise_workout_gaps as(
        select exerciseid,
            avg(log(relative_workout_gap)) as exercise_avg_workout_gap
        from
        (select distinct fake_user_id, exerciseid, workoutid, relative_workout_gap
        from fitbod a
        left join user_split b using (fake_user_id)
        left join workout_gaps d using (fake_user_id,workoutid)
        where rand <= .5) a
        group by 1 having avg(log(relative_workout_gap)) between -100 and 100)

    select exerciseid,
           exercise_avg_workout_gap,
           fake_user_id,
           avg(log(relative_workout_gap)) as user_relative_workout_gap
    from (
    select distinct fake_user_id,
           workoutid,
           d.relative_workout_gap,
           exerciseid,
           exercise_avg_workout_gap

    from fitbod a
    join exercise_workout_gaps b using (exerciseid)
    left join user_split c using (fake_user_id)
    left join workout_gaps d using (fake_user_id,workoutid)
    where rand > .5) a
    group by 1,2,3 having avg(log(relative_workout_gap)) between -100 and 100

    ''',con=engine)

    ols = sm.ols(formula = 'user_relative_workout_gap ~ exercise_avg_workout_gap', data=exercise_comparison)
    r = ols.fit()

    return {'intercept':r.params[0],'relative_workout_gap_coef':r.params[1]}

exercise_gap_projection = bootstrap_comparison(df,project_exercise_workout_gap,iterations=2000,parallel=False)


In [88]:
exercise_workout_gap_reg_df = pd.DataFrame(exercise_gap_projection)
fig = px.histogram(exercise_workout_gap_reg_df,x='relative_workout_gap_coef')
mean_coef = exercise_workout_gap_reg_df['relative_workout_gap_coef'].mean()
mean_int = exercise_workout_gap_reg_df['intercept'].mean()
print(f'The mean linear regression between log exercise relative workout gap and subsequent log relative workout gap has a coeficient of {mean_coef} and an intercept of {mean_int}')

prob_greater_than_0 = exercise_workout_gap_reg_df.loc[exercise_workout_gap_reg_df['relative_workout_gap_coef']>0]['relative_workout_gap_coef'].count()*1.0/exercise_workout_gap_reg_df['relative_workout_gap_coef'].count()
print(f'Their is an approximately a {prob_greater_than_0} probability that average exercise relative workout gap has a positive correlation with subsequent relative workout gap')
fig.show()

The mean linear regression between log exercise relative workout gap and subsequent log relative workout gap has a coeficient of 0.1287340632230837 and an intercept of -0.17495441970693193
Their is an approximately a 0.864 probability that average exercise relative workout gap has a positive correlation with subsequent relative workout gap


In [94]:
exercise_stats['rwg_projected_workout_gap'] = np.exp(-0.17495441970693193 + 0.1287340632230837*exercise_stats['avg_relative_workout_gap'])
exercise_stats['wpu_projected_workout_gap'] = np.exp(-0.1671552606740914 + 0.0006324801816318781*exercise_stats['workouts_per_user'])
exercise_stats.loc[exercise_stats['total_workouts']>=4]

exerciseid                         exercisename  \
1    mXEVvIJ5nf               Standing Hip Adduction   
4    163u4e1Em9                 Hammerstrength Shrug   
5    ISJBgnjnsY                       Single Arm Row   
6    GbK7ZSbRC6            Kettlebell Sumo High Pull   
7    JyjzR9GyUB                             TRX Pike   
8    Fa3azBAbyn                             Abs Bike   
9    olwH9a4pKv                           Knee Raise   
10   6wga8VycHO                       Clean Deadlift   
11   pM3aRwuufC        Stability Ball Hyperextension   
12   O0VUqHDOIO                     TRX Swimmer Pull   
13   Yw1aA3soIp                  Dumbbell Sumo Squat   
14   r7LqFtCsNv                    Decline Leg Raise   
15   AOQBAQ8rXC                          Pendlay Row   
16   iQzrqsifIr                      Lateral Step Up   
17   crpeAKEXTd                   TRX Atomic Push Up   
18   y2Gg0PEzd8                        Preacher Curl   
19   QUvE80fdlJ                      Tricep Kickback   
20   3GvUTFN8up                           Calf Raise   
21   C4tOdXyE0R                  Cable Hip Abduction   
22   wNImc4b4MY                             Leg Curl   
23   qfkBJ4A5eQ                             Dead Bug   
24   1sQqqUqJyp                            Frog Pump   
25   jgnSb5e39S                   Australian Chin Up   
26   so3n0BqhED             Smith Machine Calf Raise   
27   XlazCsjHwn                          Hang Snatch   
28   ewEcdHovom                    Hanging Leg Raise   
29   yR61QnjQAO                Medicine Ball Push Up   
30   BN1DjSE1Zi                    Romanian Deadlift   
31   58REE5sJU1                        Russian Twist   
32   erjFOF21dc                 Reverse Barbell Curl   
33   QFhUumXlvb   Reverse Grip Barbell Bent Over Row   
34   243X43p1XP                      Seated Leg Curl   
35   9fptvki4pW                   Barbell Hip Thrust   
36   7wof3tzAVY  Smith Machine Stiff-Legged Deadlift   
37   zJlvUiMLou          Barbell Incline Bench Press   
38   k3yxx2a6oz            Seated Machine Calf Press   
39   06kzIMnTzR                       Oblique Crunch   
40   nNMm3ywt6H                         Pike Push Up   
41   516XLLabVp                     Scapular Pull Up   
42   LDHzRzwxaI                Exercise Ball Push Up   
43   otjt26R0oL               Barbell Shoulder Press   
44   aIzmfLklht                     Dumbbell Step Up   
45   gALxFFZcI4         Dumbbell Decline Bench Press   
46   Mok6zwcCl9                    Cable Upright Row   
47   uHBREgyju0                              Step Up   
48   k4ivWt2zuv                 Barbell Glute Bridge   
49   oDI8CKhqJi             Single Arm Preacher Curl   
50   9q2Cdgm0FM                    Machine Leg Press   
51   HpJxPvuCBP                       Dumbbell Clean   
52   G1mxtZtEs5                           Back Squat   
53   UThEK7cf4f                          Machine Fly   
54   DhkjvJNsNR                               Sit Up   
55   DdtF1OFcuQ                       Tricep Push Up   
56   UQBXLgsfXN                       Dumbbell Shrug   
57   r4bOO0fDTv                                Plank   
58   WoEeZr6Uwx              Single Arm Lat Pulldown   
59   mIMgSxPI0T                          Toes to Bar   
60   ACSgVaipTW               Dumbbell Squeeze Press   
61   BTmBzxhAmZ                    Dumbbell Pullover   
62   fxXbcxfkLt                 Exercise Ball Crunch   
63   dFycDkJyHe              Dumbbell Shoulder Raise   
64   4IMiKZHvCG                          Shotgun Row   
65   5hsZsuO5vt                Single Leg Hip Thrust   
66   JZcl0pTDEj                     TRX Glute Bridge   
67   9VAkz3bNl4                      Decline Push Up   
68   HggeHLeipS              Cable Double Bicep Curl   
69   EAjKaB5zWy                   Dumbbell Kickbacks   
70   KVt5QRq9Hx         Side Laterals to Front Raise   
71   PK3FADlv5b                          Floor Press   
72   vWKV6qv8Pv                       Clean and Jerk   
73   wNarovWdLy              

## Discussion

The relationship between exercise predicted relative workout gap and actual relative workout gap shows some promise here. The 87% probability of a positive relationship is promising given the small sample size, especially given that the intention is not neccessarily to use this metric as a predictor itself.

What is most important here is that we are at least reasonably sure that this measure has consistentcy across users, which tells us that the metric is not just noise and shows some reflection of user behavior in response to specific exercises that is similar across users. On top of consistency of this metric across users, this metric's predictive nature also points towards some consistency in the context in which these workouts are given to users, which is good to know. 

## Next Steps

While we will likely never use these metrics to directly recommend exercises in practice, it can be helpful to understand how big the effect size would be were we to optimize to this variable and what issues would arrise. Lets look at what recommending workouts based on exercise usage rate and exercise relative workout gap might look like in practice to understand that potential effect size and analyze what possible contributing factors we can look into next.

If we take a legs workout with 3 legs and 2 abs exercises and migrate it from the most common workouts to the low end of the specturm of exercise-level predicted workout gaps, it might look somehting like this:

#### Original workout:
- Air Squats --> 
- Back Squats 
- Seated Leg Curl 
- Plank 
- Leg Raise 

Avg predicted relative workout gap: 0.886

#### Low relative workout gap workout:
- Dumbbell Sumo Squat	(0.785) --> exp(-0.175 + 0.1287*log(0.785)) = 0.841
- Standing Hip Adduction (0.489) --> exp(-0.175 + 0.1287*log(0.489)) = 0.765
- Calf Raise (0.812) --> exp(-0.175 + 0.1287*log(0.812)) = 0.846
- TRX Pike (0.676) --> exp(-0.175 + 0.1287*log(0.676)) = 0.816
- Abs Bike (0.676) --> exp(-0.175 + 0.1287*log(0.676)) = 0.816

Avg predicted relative workout gap: 0.817

#### Differences in subsequent workout gap
.886 - .817 = .069

For sake of comparison, lets compare the workout and the predicted outcome to a workout with exercises with low workout frequency and its predicted outcome. We know it isn't a highly likely relationship, but understanding the magnitude could be informative of where further inquiry could be directed.

First lets pull up the lowest workouts per user exercises and create a workout from there.



In [95]:
exercise_stats.loc[exercise_stats['total_workouts']>=4].sort_values('workouts_per_user')

exerciseid                         exercisename  \
16   iQzrqsifIr                      Lateral Step Up   
93   PZ2ogZtiY4                     TRX Pulse Lunges   
98   tSjey8z21h                      TRX Chest Press   
66   JZcl0pTDEj                     TRX Glute Bridge   
6    GbK7ZSbRC6            Kettlebell Sumo High Pull   
1    mXEVvIJ5nf               Standing Hip Adduction   
29   yR61QnjQAO                Medicine Ball Push Up   
8    Fa3azBAbyn                             Abs Bike   
14   r7LqFtCsNv                    Decline Leg Raise   
72   vWKV6qv8Pv                       Clean and Jerk   
5    ISJBgnjnsY                       Single Arm Row   
12   O0VUqHDOIO                     TRX Swimmer Pull   
42   LDHzRzwxaI                Exercise Ball Push Up   
95   jJmv4Zv028             Overhead Weighted Sit Up   
7    JyjzR9GyUB                             TRX Pike   
99   LaUm5Jl8hd                            Sled Pull   
26   so3n0BqhED             Smith Machine Calf Raise   
71   PK3FADlv5b                          Floor Press   
11   pM3aRwuufC        Stability Ball Hyperextension   
10   6wga8VycHO                       Clean Deadlift   
76   AkKyo5MAvA               Kettlebell Front Squat   
44   aIzmfLklht                     Dumbbell Step Up   
100  geCUwfcZLR                           Hang Clean   
21   C4tOdXyE0R                  Cable Hip Abduction   
84   Gx6Z2fevPD                 Incline Dumbbell Row   
90   R1M0DI9wui          Standing Hip Flexor Stretch   
97   9VQ4n2PTFW                       Landmine Press   
9    olwH9a4pKv                           Knee Raise   
15   AOQBAQ8rXC                          Pendlay Row   
74   cDYgx7gYHz                           Tate Press   
45   gALxFFZcI4         Dumbbell Decline Bench Press   
92   RqCbxJr56l                         Tricep Press   
65   5hsZsuO5vt                Single Leg Hip Thrust   
87   oySTEjJdfw               Freemotion Chest Press   
91   w5njIJMJHb               Kettlebell Upright Row   
62   fxXbcxfkLt                 Exercise Ball Crunch   
28   ewEcdHovom                    Hanging Leg Raise   
27   XlazCsjHwn                          Hang Snatch   
89   s5tzCNPRXw                           Lunge Jump   
13   Yw1aA3soIp                  Dumbbell Sumo Squat   
25   jgnSb5e39S                   Australian Chin Up   
49   oDI8CKhqJi             Single Arm Preacher Curl   
88   9ykyHFs36M                                Clean   
48   k4ivWt2zuv                 Barbell Glute Bridge   
54   DhkjvJNsNR                               Sit Up   
39   06kzIMnTzR                       Oblique Crunch   
51   HpJxPvuCBP                       Dumbbell Clean   
17   crpeAKEXTd                   TRX Atomic Push Up   
94   tTFy5YOe7N            Single Arm Overhead Press   
40   nNMm3ywt6H                         Pike Push Up   
78   omtypRFVyt                          Power Clean   
77   dOrcZrFXST                  Handle Band Pec Fly   
4    163u4e1Em9                 Hammerstrength Shrug   
63   dFycDkJyHe              Dumbbell Shoulder Raise   
59   mIMgSxPI0T                          Toes to Bar   
73   wNarovWdLy                         Cable Crunch   
30   BN1DjSE1Zi                    Romanian Deadlift   
18   y2Gg0PEzd8                        Preacher Curl   
57   r4bOO0fDTv                                Plank   
70   KVt5QRq9Hx         Side Laterals to Front Raise   
32   erjFOF21dc                 Reverse Barbell Curl   
36   7wof3tzAVY  Smith Machine Stiff-Legged Deadlift   
35   9fptvki4pW                   Barbell Hip Thrust   
23   qfkBJ4A5eQ                             Dead Bug   
67   9VAkz3bNl4                      Decline Push Up   
82   AF7Qq7XK2r                           Air Squats   
47   uHBREgyju0                              Step Up   
81   N6zu9IJNBz                            Leg Raise   
33   QFhUumXlvb   Reverse Grip Barbell Bent Over Row   
64   4IMiKZHvCG                          Shotgun Row   
24   1sQqqUqJyp              

#### Original workout:
- Air Squats (4.717) --> exp(-0.122 + 0.000277*4.717) = .886
- Back Squats (10.279) --> exp(-0.122 + 0.000277*10.279) = .888
- Seated Leg Curl (8.167) --> exp(-0.122 + 0.000277*8.167) = .887
- Plank (4.327) --> exp(-0.122 + 0.000277*4.327) = .886
- Leg Raise (4.820) --> exp(-0.122 + 0.000277*4.820) = .886

Avg exercise relative workout gap: .887

#### Low exercise workout frequncy workout:
- Kettlebell Front Squat (2.333) --> exp(-0.122 + 0.000277*2.333) = .886
- TRX Pulse Lunges (1.125) --> exp(-0.122 + 0.000277*1.125) = .885
- Lateral Step Up (1.250) --> exp(-0.122 + 0.000277*1.250) = .885
- Decline Leg Raise (1.667) --> exp(-0.122 + 0.000277*1.667) = .885
- Abs Bike (1.571) --> exp(-0.122 + 0.000277*1.571) = .885

Avg exercise workout frequency: .885

#### Differences in subsequent workout gap
.887 - .885 = .002


## Discussion

When comparing the two workouts, we see both a much higher probability of positive correlation (84.2% vs 55.9%) and a much larger predicted effect size (.069 vs .002) using the exercise relative workout gap input instead of the exercise workout frequency. 

While we could potentially use the measure of exercise relative workout gaps to help inform workout curation, using this metric as an input could cause problems. First of all, many of the same exercises that would be chosen by frequency would also be chosen by workout gap, so it is uncertain whether the measurements are hitting on dramatically different signals. Additionally, the measurement is something of a catch-all of reaction to an exercise that ignores the surrounding context of an exercise.

What both of these measurements have in common is that they are measurements that are likely to have underlying compontent parts. If we can locate one or more of these component parts by looking at what these measurements might be indirectly indiating, then we may be able to find a direct action that could improve the fitbod user experience. 

## Next Steps

I want to break down exercise workout frequency into possible compontent parts that could be more direct measures of a variable that is causal in increasing user activity, so I am going to start with exercise number. If less frequented workouts tend to produce shorter subsequent workout gaps, that could potentially be because the workout is more likely to be fresh to a user

In [44]:
exercise_new = pd.read_sql('''
select case when exercise_num = 1 then 'new' else 'previously performed' end as new_exercise,
       avg(relative_workout_gap)) as avg_log_rel_workout_gap,
       stddev(relative_workout_gap)/(count(*))^.5 as std_err_rel_workout_gap,
       count(*) as sample_size,
       avg(avg_workout_num) as avg_workout_num

from
(Select fake_user_id,
        exercise_num,
        avg(log(relative_workout_gap)) as relative_workout_gap,
        avg(workout_num) as avg_workout_num
from
(Select distinct fake_user_id, 
        workoutid,
        exerciseid,
        b.relative_workout_gap,
        c.workouts_per_user,
        dense_rank() over (partition by fake_user_id,exerciseid order by date(date)) as exercise_num,
        dense_rank() over (partition by fake_user_id order by date(date)) as workout_num
from fitbod a
left join workout_gaps b using (fake_user_id,workoutid)
left join fitbod_exercise_stats c using (exerciseid)) a
where workout_num >= 3
group by 1,2) a
group by 1 having count(*) > 10
''',con=engine)
exercise_new.head(10)

new_exercise  avg_rel_workout_gap  std_err_rel_workout_gap  \
0  previously performed             0.879031                 0.007478   
1                   new             0.783394                 0.018911   

   sample_size  avg_workout_num  
0         1144        77.901887  
1           68        22.131019

In [18]:
change = round((1.096104-0.873289)*100/1.096104,2)
t = (1.096104-0.873289)/(0.034009**2 + 0.027794**2)**.5
p = scipy.stats.t.sf(t,df=67)
print(f'users have a {change}% shorter relative workout gap when the workout includes a first time exercise vs not with a t-value of {t} and a p-value of {p}')
print()

users have a 20.33% shorter relative workout gap when the workout includes a first time exercise vs not with a t-value of 5.073001252246189 and a p-value of 1.660713494788659e-06



The presence of a net-new workout has a statistically significant effect on subsequent workout gap, with net-new exercises prompting a 20.33% shorter relative workout gap than non-net-new exercises. With this relationship seemingly clear in the data, I would love to understand a bit more about how much exercise newness is optimal in terms of producing engageing workouts that keep users coming back next time. This question also has the added benefit of dividing out workouts by the presenece of new exercises, whereas my previous measurement has some overlap because it only looked at exercise workout number and not the combination of exercises.

In [19]:
exercise_new = pd.read_sql('''
select number_first_time_exercises,
       avg(relative_workout_gap) as avg_relative_workout_gap,
       stddev(relative_workout_gap)/(count(*))^.5 as stderr_rel_workout_gap,
       count(*) as sample_size
from (
        select fake_user_id,
               number_first_time_exercises,
               avg(relative_workout_gap) as relative_workout_gap
        from (
                Select fake_user_id,
                        workoutid,
                        workout_num,
                        relative_workout_gap,
                        count(distinct case when exercise_num = 1 then exerciseid end) as number_first_time_exercises
                from (
                        Select distinct fake_user_id, 
                                workoutid,
                                exerciseid,
                                b.relative_workout_gap,
                                c.workouts_per_user,
                                dense_rank() over (partition by fake_user_id,exerciseid order by date(date)) as exercise_num,
                                dense_rank() over (partition by fake_user_id order by date(date)) as workout_num
                        from fitbod a
                        left join workout_gaps b using (fake_user_id,workoutid)
                        left join fitbod_exercise_stats c using (exerciseid)) a
                where workout_num >= 4
                group by 1,2,3,4) b
        group by 1,2
) c
group by 1 --having count(*) > 10
order by 1
''',con=engine)
exercise_new.head(10)

number_first_time_exercises  avg_relative_workout_gap  \
0                            0                  1.097553   
1                            1                  0.907597   
2                            2                  0.914581   
3                            3                  0.838059   
4                            4                  0.807264   
5                            5                  0.378271   

   stderr_rel_workout_gap  sample_size  
0                0.040450           65  
1                0.036836           67  
2                0.132023           59  
3                0.130823           32  
4                0.213783           16  
5                0.038648            2

Well isn't that pretty! While the standard errors are fairly high for the upper end of the specturm, there look to be strong beneficial user responses to including greater amounts of net-new exercises in workouts. While this relationship seems solid on the surface, it is still possible that this effect could be at least paritally due to trends in workout number. Because a user is more likely to recieve an exercise for the first time in earlier workouts in their fitbod tenure, if workout gaps are shorter earlier in a user's fitbod tenure, then that relationship could produce the trend in net-new exercises we see above. Even if we prove that earlier workouts tend to have shorter relative gaps, it is tough to know for certain the direction of causality between early tenure and net-new exercises, as either one could cause the other. 

All of this is moot if there is no relationship between workout number and relative workout gap, so lets check in to see if that relationship exists. To do so, I am going to run a linear regression between the workout number and the relative workout gap to see if relative workout gap increases over fitbod usage tenure. Because I am somewhat excited by new exercies driving engagement, I am going to steelman the opposing argument and see if I can find the best opposing representation of the data to suggest that workout number is the main driver of this statistic.

In [20]:
workout_num_relative_gap = pd.read_sql('''
select workout_num,
       log(avg(relative_workout_gap)) as log_relative_workout_gap,
       stddev(relative_workout_gap)/(count(distinct workoutid))^.5 as stderr_rel_workout_gap,
       count(distinct workoutid) as sample_size
from
(select fake_user_id,
       workoutid,
       relative_workout_gap,
       workout_date,
       row_number() over (partition by fake_user_id order by workout_date,workoutid) as workout_num
from 
(Select distinct fake_user_id, 
        workoutid,
        b.relative_workout_gap,
        date(date) as workout_date
from fitbod a
left join workout_gaps b using (fake_user_id,workoutid) ) a ) b
group by 1 having count(*) > 15
order by 1
''',con=engine)
workout_num_relative_gap.head(20)

workout_num  log_relative_workout_gap  stderr_rel_workout_gap  sample_size
0             1                  0.000470                0.124690           71
1             2                 -0.099021                0.089536           69
2             3                 -0.134866                0.079414           69
3             4                 -0.129575                0.074819           68
4             5                 -0.148666                0.060812           67
5             6                 -0.123558                0.075189           67
6             7                 -0.083670                0.103060           67
7             8                 -0.017589                0.237059           66
8             9                 -0.169986                0.077038           66
9            10                  0.093770                0.282441           65
10           11                 -0.085320                0.095625           64
11           12                 -0.050420                0.108576           64
12           13                 -0.077820                0.103351           64
13           14                  0.148329                0.302465           62
14           15                  0.075868                0.194128           62
15           16                 -0.071612                0.212344           61
16           17                 -0.183709                0.077103           61
17           18                  0.051300                0.187715           60
18           19                 -0.093375                0.097260           60
19           20                 -0.001573                0.177327           60

In [21]:
fig = px.scatter(workout_num_relative_gap, x="workout_num", y="log_relative_workout_gap", trendline="ols")
fig.show()

results = px.get_trendline_results(fig)
print(results)

results.px_fit_results.iloc[0].summary()

                                      px_fit_results
0  <statsmodels.regression.linear_model.Regressio...


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.048
Model:                            OLS   Adj. R-squared:                  0.039
Method:                 Least Squares   F-statistic:                     5.459
Date:                Thu, 07 Apr 2022   Prob (F-statistic):             0.0213
Time:                        09:42:03   Log-Likelihood:                 113.95
No. Observations:                 111   AIC:                            -223.9
Df Residuals:                     109   BIC:                            -218.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0395      0.017     -2.364      0.020      -0.073      -0.006
x1             0.0006      0.000      2.337      0.021    9.19e-05       0.001
==============================================================================
Omnibus:                        6.112   Durbin-Watson:                   1.805
Prob(Omnibus):                  0.047   Jarque-Bera (JB):                6.242
Skew:                           0.380   Prob(JB):                       0.0441
Kurtosis:                       3.879   Cond. No.                         130.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

There does look to be a relationship between workout number and relative workout gap, albeit something of a weak one. Every representation of data has its positives and negatives. This representation of the data takes an average of all users at a given workout number. This attempts to incorporate all workout numbers equally in the cration of the equation in spite of more users performing earlier workout numbers than later ones. Lets try allocating a data point for each pair of user and workout number instead and see how it changes the resulting relationship.

In [22]:
workout_num_relative_gap_all_points = pd.read_sql('''
select fake_user_id,
       workoutid,
       log(relative_workout_gap) as log_rel_workout_gap,
       workout_date,
       row_number() over (partition by fake_user_id order by workout_date,workoutid) as workout_num
from 
(Select distinct fake_user_id, 
        workoutid,
        b.relative_workout_gap,
        date(date) as workout_date
from fitbod a
left join workout_gaps b using (fake_user_id,workoutid) 
where relative_workout_gap <> 0) a 
''',con=engine)
workout_num_relative_gap_all_points.head(20)

fake_user_id   workoutid  log_rel_workout_gap workout_date  workout_num
0   fake_id_433  BQCCG4RTqN            -0.222353   2021-03-08            1
1   fake_id_433  nWLnpEuu28            -0.222353   2021-03-09            2
2   fake_id_433  9FJxo0E4i8            -0.222353   2021-03-10            3
3   fake_id_433  xLz1NKGPKC             0.078677   2021-03-11            4
4   fake_id_433  Gbi0D7fe3g             0.254768   2021-03-13            5
5   fake_id_433  b1elTeBkem            -0.222353   2021-03-16            6
6   fake_id_433  VvnsAp4MSv             0.476617   2021-03-17            7
7   fake_id_433  lnKq6MR7QU            -0.222353   2021-03-22            8
8   fake_id_433  ZQNg1D8BZL            -0.222353   2021-03-23            9
9   fake_id_433  9QOjhXBUGq             0.078677   2021-03-24           10
10  fake_id_433  6UrUyGtMzB             0.379707   2021-03-26           11
11  fake_id_433  UEEvLXuSBV             0.078677   2021-03-30           12
12  fake_id_433  Zqkl3nwIlv             0.254768   2021-04-01           13
13  fake_id_433  P1xSywiqSi             0.680737   2021-04-04           14
14  fake_id_433  CQXL4YiRNR             0.078677   2021-04-12           15
15  fake_id_433  piRPRZDYN8             0.254768   2021-04-14           16
16  fake_id_433  RY0YBC4zQk             0.078677   2021-04-17           17
17  fake_id_433  aGNNtxxwrg             0.078677   2021-04-19           18
18  fake_id_433  eSqeIze6vn            -0.222353   2021-04-21           19
19  fake_id_433  4Eqz0YPxBu             0.379707   2021-04-22           20

In [23]:
fig = px.scatter(workout_num_relative_gap_all_points.loc[workout_num_relative_gap_all_points['workout_num'] <= 60], x="workout_num", y="log_rel_workout_gap", trendline="ols")
fig.show()

results = px.get_trendline_results(fig)
print(results)

results.px_fit_results.iloc[0].summary()

                                      px_fit_results
0  <statsmodels.regression.linear_model.Regressio...


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.015
Model:                            OLS   Adj. R-squared:                  0.015
Method:                 Least Squares   F-statistic:                     45.86
Date:                Thu, 07 Apr 2022   Prob (F-statistic):           1.51e-11
Time:                        09:42:03   Log-Likelihood:                -1225.1
No. Observations:                3047   AIC:                             2454.
Df Residuals:                    3045   BIC:                             2466.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.2383      0.012    -19.439      0.000      -0.262      -0.214
x1             0.0026      0.000      6.772      0.000       0.002       0.003
==============================================================================
Omnibus:                       97.161   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              108.898
Skew:                           0.418   Prob(JB):                     2.25e-24
Kurtosis:                       3.398   Cond. No.                         59.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [24]:
fig = px.scatter(workout_num_relative_gap_all_points.loc[workout_num_relative_gap_all_points['workout_num'] > 60], x="workout_num", y="log_rel_workout_gap", trendline="ols")
fig.show()

results = px.get_trendline_results(fig)
print(results)

results.px_fit_results.iloc[0].summary()

                                      px_fit_results
0  <statsmodels.regression.linear_model.Regressio...


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.009
Model:                            OLS   Adj. R-squared:                  0.009
Method:                 Least Squares   F-statistic:                     16.08
Date:                Thu, 07 Apr 2022   Prob (F-statistic):           6.34e-05
Time:                        09:42:03   Log-Likelihood:                -272.40
No. Observations:                1720   AIC:                             548.8
Df Residuals:                    1718   BIC:                             559.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1121      0.018     -6.120      0.000      -0.148      -0.076
x1             0.0006      0.000      4.010      0.000       0.000       0.001
==============================================================================
Omnibus:                      121.008   Durbin-Watson:                   1.996
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.539
Skew:                           0.644   Prob(JB):                     1.24e-33
Kurtosis:                       3.674   Cond. No.                         325.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

So there is indeed a relationship between workout number and relative workout gap. I wound up splitting the regression into two because it seemed as though the high end of workout number had a narrower range of relative workout gaps, which makes sense given that users who worked out enough to make it up to that point have inherently fewer skip days in their schedule, making the variability of their gaps lower. 

Lets see if the relationship between workout number and relative workout gap is enough to explain the differences caused by users performing net-new exercises. To do this, I am going to perform the same calculation of net-new exercise impact and include the expected workout gap given the workout number. I will be using the strongest relationship between workout number and workout gap which was acheived using the individual user representation. 

In [25]:
exercise_new_workout_adjusted = pd.read_sql('''
select number_first_time_exercises,
       avg(relative_workout_gap) as avg_relative_workout_gap,
       stddev(relative_workout_gap)/(count(*))^.5 as stderr_rel_workout_gap,
       avg(workout_num_expected_rel_gap) as workout_num_expected_rel_gap,
       stddev(workout_num_expected_rel_gap)/(count(*))^.5 as stderr_workout_num_expected_rel_gap,
       count(*) as sample_size
from (
        select fake_user_id,
               number_first_time_exercises,
               avg(relative_workout_gap) as relative_workout_gap,
               avg(workout_num_expected_rel_gap) as workout_num_expected_rel_gap
        from (
                Select fake_user_id,
                        workoutid,
                        workout_num,
                        relative_workout_gap,
                        case when workout_num < 60 then exp(-0.2383 + .0026*workout_num) else exp(-0.1121 + .0006*workout_num) end as workout_num_expected_rel_gap,
                        count(distinct case when exercise_num = 1 then exerciseid end) as number_first_time_exercises
                from (
                        Select distinct fake_user_id, 
                                workoutid,
                                exerciseid,
                                b.relative_workout_gap,
                                c.workouts_per_user,
                                dense_rank() over (partition by fake_user_id,exerciseid order by date(date),workoutid) as exercise_num,
                                dense_rank() over (partition by fake_user_id order by date(date),workoutid) as workout_num
                        from fitbod a
                        join workout_gaps b using (fake_user_id,workoutid)
                        join fitbod_exercise_stats c using (exerciseid)
                        where relative_workout_gap is not null) a
                where workout_num >= 4
                group by 1,2,3,4,5) b
        group by 1,2
) c
group by 1 --having count(*) > 15
order by 1
''',con=engine)
exercise_new_workout_adjusted.head(10)

number_first_time_exercises  avg_relative_workout_gap  \
0                            0                  1.100763   
1                            1                  0.906175   
2                            2                  0.911660   
3                            3                  0.844080   
4                            4                  0.807264   
5                            5                  0.378271   

   stderr_rel_workout_gap  workout_num_expected_rel_gap  \
0                0.039515                      0.871088   
1                0.038799                      0.844995   
2                0.132355                      0.824357   
3                0.130001                      0.819850   
4                0.213783                      0.802716   
5                0.038648                      0.802441   

   stderr_workout_num_expected_rel_gap  sample_size  
0                             0.005410           65  
1                             0.003996           66  
2                             0.003661           59  
3                             0.007135           32  
4                             0.001335           16  
5                             0.002086            2

While the workout number prediction did mirror that of the first-time exercises, the maginitude of the predicted difference was lesser than the difference showed by the net new workouts. This strengthens the validity of the relationship between new exercises and improved engagement. 

Next, I want to test if the net-new exercieses are actually explanatory of all of the relationship in terms of workout number. To do so, I am going to take the average relative workout gap predicted by the number of first time exercises and look to see if a postive linear relationship still exists in the difference between the predicted and actual workout gaps.

In [26]:
exercise_new_workout_adjusted.to_sql('number_first_time_exercises',con=engine, if_exists='replace') 
workout_number_new_exercise_adjusted = pd.read_sql('''
select workout_num,
       avg(relative_workout_gap - coalesce(c.avg_relative_workout_gap,1)) as net_workout_gap
from (
        Select fake_user_id,
                workoutid,
                workout_num,
                relative_workout_gap,
                count(distinct case when exercise_num = 1 then exerciseid end) as number_first_time_exercises
        from (
                Select distinct fake_user_id, 
                        workoutid,
                        exerciseid,
                        b.relative_workout_gap,
                        c.workouts_per_user,
                        dense_rank() over (partition by fake_user_id,exerciseid order by date(date),workoutid) as exercise_num,
                        dense_rank() over (partition by fake_user_id order by date(date),workoutid) as workout_num
                from fitbod a
                join workout_gaps b using (fake_user_id,workoutid)
                join fitbod_exercise_stats c using (exerciseid)
                where relative_workout_gap is not null) a
        where workout_num >= 4
        group by 1,2,3,4) b
left join number_first_time_exercises c using (number_first_time_exercises)
group by 1 having count(*) > 15
''',con=engine)

In [27]:
fig = px.scatter(workout_number_new_exercise_adjusted, x="workout_num", y="net_workout_gap", trendline="ols")
fig.show()

results = px.get_trendline_results(fig)
print(results)

results.px_fit_results.iloc[0].summary()

                                      px_fit_results
0  <statsmodels.regression.linear_model.Regressio...


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.7633
Date:                Thu, 07 Apr 2022   Prob (F-statistic):              0.384
Time:                        09:42:04   Log-Likelihood:                 14.538
No. Observations:                 107   AIC:                            -25.08
Df Residuals:                     105   BIC:                            -19.73
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0741      0.043     -1.713      0.090      -0.160       0.012
x1             0.0006      0.001      0.874      0.384      -0.001       0.002
==============================================================================
Omnibus:                       50.764   Durbin-Watson:                   1.941
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              187.523
Skew:                           1.608   Prob(JB):                     1.90e-41
Kurtosis:                       8.632   Cond. No.                         136.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Discussion

While the relationship between workout number and relative workout gap did not go away entirely, the significance of the relationship shrunk to less than half of the previous value when comparing to the same linear regression from earlier. If I was looking at the linear relationship between these two varaibles without any surrounding context, I would assume there was no relationship between the two variables. Given the results of the regression when adjusting for number of first time exercises, I think it is now safe to conclude that including first-time exercises in a workout reduces the relative workout gap following the exercise. 

If I had the data on hand, I would be very interested to know how often these first-time exercises were the result of user intervention and how often they were the result of fitbod reccomendation. I would also be interested to know how often users, when presented with a new exercise, elected to replace that exercise with one that was known to them.

## Next Steps

### Designing an A/B test

For the first time in this analysis, we have a relationship that clearly suggests causality. While the correlation of these two variables looks like causality, we still need to run a test to confirm that there is not some other unforseen driver of this relationship. 

To test that exercise novelty is driving increased engagement, I would run an A/B test where users are divided into one control and two test groups after their third workouts. In the first test group, users would be hard-coded to recieve at least one net-new exercise per workout. In the second test group, users would recieve a boost to the probability of a net-new workout but would not be hard-coded to recieve it (I would work with those more familiar with the algorithm to understand how this could be optimally implemented). The control group would receive no changes to their fitbod exeprience. I would also present an out from the test groups whereby users who replaced 3 first-time exercises with previously performed exercises would be places back in the control group to prevent attrition of users who did not enjoy the intervention.

To evaluate the success of this test, the primary target metrics would be workout gap and relative workout gap of users, with decreases to workout gap of course being positive. Relative workout gap would be more helpful in terms of determining the impact of this change on longer-tenured users, while non-relative workout gap would be helpful in comparing early-stage users at the conclusion of this test. I would also be interested to compare the attrition rate of the users in each of the groups as well as the rates at which users replaces the net-new exercises and the replacements they chose. 

The hope would be that this test would illuminate not only if adding emphasis on net-new exercises was beneficial in practice, but also if there were diminishing returns to implementing this kind of variety consistently over time. The time-dependent nature of the change would be quite important when determining the success of this test, as novelty could quite easily be overdone and lose its potency over time. Understanding decay curve of novelty impact could additionally be quite helpful in effectively implementing changes to the workout recommendation algorithm. 

### Analysis Next Steps

The impact of exercise novelty in user engagement makes me wonder about a related avenue that could drive user engagement. 